akka-typed/src/main/scala/akka/typed/cluster/ClusterSingleton.scala

```diff
@@ -0,0 +1,218 @@
+  override def createExtension(system: ActorSystem[_]): ClusterSingleton = new ClusterSingletonImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object ClusterSingletonImpl {
+  def managerNameFor(singletonName: String) = s"singleton-manager-${singletonName}"
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] final class ClusterSingletonImpl(system: ActorSystem[_]) extends ClusterSingleton {
+  require(system.isInstanceOf[ActorSystemAdapter[_]], "only adapted actor systems can be used for the typed cluster singleton")
+  import ClusterSingletonImpl._
+
+  private lazy val cluster = Cluster(system)
+  private val untypedSystem = ActorSystemAdapter.toUntyped(system).asInstanceOf[ExtendedActorSystem]
+
+  override def spawn[A](
+    behavior:           Behavior[A],
+    singletonName:      String,
+    props:              Props,
+    settings:           ClusterSingletonSettings,
+    terminationMessage: A) = {
+
+    spawnManager(behavior, singletonName, props, settings, terminationMessage)
+    spawnProxy(singletonName, Props.empty, settings)
+  }
+
+  def spawnManager[A](behavior: Behavior[A], singletonName: String, props: Props, settings: ClusterSingletonSettings, terminationMessage: A): Boolean = {
+    if (settings.shouldRunManager(cluster)) {
+      val managerName = managerNameFor(singletonName)
+      // start singleton on this node
+      val adaptedProps = PropsAdapter(behavior, props)
+      try {
+        untypedSystem.systemActorOf(
+          OldSingletonManager.props(adaptedProps, terminationMessage, settings.toManagerSettings(singletonName)),
+          managerName)
+      } catch {
+        case ex: InvalidActorNameException ⇒
+        // This is fine. We just wanted to make sure it is running and it already is
+      }
+      true
+    } else
+      false
+
+  }
+
+  def spawnProxy[A](singletonName: String, props: Props, settings: ClusterSingletonSettings): ActorRef[A] = {
+    val proxyName = s"singleton-proxy-${singletonName}"
+    val untypedProxy =
+      try {
+        untypedSystem.systemActorOf(
+          ClusterSingletonProxy.props(s"/system/${managerNameFor(singletonName)}", settings.toProxySettings(singletonName)),
+          proxyName)
+      } catch {
+        case ex: InvalidActorNameException ⇒
+          // this is fine, we don't want to start more of it than one
+          untypedSystem.actorFor(s"/system/$proxyName")
+      }
+    ActorRefAdapter[A](untypedProxy)
+  }
+}
+
+/**
+ * Not intended for user extension.
+ */
+@DoNotInherit
+trait ClusterSingleton extends Extension {
+
+  // Ideas:
+  //  1. single method creates both manager and proxy
+  //  2. method signature mimics the general actor starting signature and name
+  // if needed we could also provide a spawnManager and a spawnProxy (but I think those are only ever used
+  // to do the same kind of if-else-with-role/dc in user logic)
+
+  // TODO what happens if there already is a singletonmgr for the name running?
```

Comment:
Alternatives are to either throw the name not unique thing or return the existing one...
The latter could be racy, and the prior is expected anyway - so just throw I'd say, we could enrich the exception with a more informative message though - "hey plz start singleton only once on a given node"